### PR TITLE
Adds query parser and middleware for get task requests

### PR DESCRIPTION
**Pull request by Ajeyakrishna-k**
Date: 14 November 2023
Developer Name: @Ajeyakrishna-k 

----
## Issue Ticket Number:- 
-  #1588 

## Description: 
- Adds a query parser function to parse query params based on RDS query language format. [Docs](https://github.com/Real-Dev-Squad/docs/pull/20)
- Adds a middleware validator for GET /task-requests route 

Is Under Feature Flag 
- [x] Yes
- [ ] No

Database changes
- [ ] Yes
- [x] No

Breaking changes (If your feature is breaking/missing something please mention pending tickets)
- [ ] Yes
- [x] No

### Is Development Tested?

- [x] Yes
- [ ] No

Test code coverage: 
Task requests validator: 
Screenshot 2023-11-14 at 4 01 11 PM

RQL query params parser: 
Screenshot 2023-11-14 at 4 00 45 PM

### CODE REVIEW

File: middlewares/validators/task-requests.js [hunk 1]
```diff
@@ -40,6 +41,61 @@ const postTaskRequests = async (req, res, next) => {
...
+    .keys({
+      dev: joi.bool().optional().sensitive(),
+      prev: joi.string().optional(),
+      next: joi.string().optional(),
+      size: joi.number().integer().positive().min(1).max(100).optional(),
+      q: joi.string().optional(),
+    })
+    .without("prev", "next")
+    .with("prev", "size")
+    .with("next", "size");
```
Comment (shubhamsigdar1) [inline middlewares/validators/task-requests.js:56]: Nit: 
```
.without("prev", "next")
.with("prev", "size")
.with("next", "size");
```

What do you mean by above 3 line?

File: middlewares/validators/task-requests.js [hunk 2]
```diff
@@ -40,6 +41,61 @@ const postTaskRequests = async (req, res, next) => {
   }
 };
 
+const getTaskRequests = async (req, res, next) => {
+  const schema = joi
+    .object()
```
Comment (shubhamsigdar1) [inline middlewares/validators/task-requests.js:46]: Nit: This look like pagination schema

File: middlewares/validators/task-requests.js (hunk 1 above)
Comment (Ajeyakrishna-k) [reply middlewares/validators/task-requests.js:56]: https://joi.dev/api/?v=17.9.1#objectwithkey-peers-options

File: middlewares/validators/task-requests.js (hunk 2 above)
Comment (Ajeyakrishna-k) [reply middlewares/validators/task-requests.js:46]: Changed it to `queryParamSchema`

File: utils/RQLParser.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+interface GroupedSortQueries {
+  [key: string]: string;
+}
+
+const KEY_VALUE_PAIR_DELIMITER_PATTERN = /\s+/;
+const KEY_VALUE_DELIMITER = ":";
+const SORT_DELIMITER = "-";
+const EXCLUDE_SYMBOL = "-";
```
[the remark this answers is not in the document]
Comment (Ajeyakrishna-k) [reply utils/RQLParser.ts:34]: Can't exclude a sort. So these two can't be together.

File: middlewares/validators/task-requests.js (hunk 1 above)
Comment (shubhamsigdar1) [reply middlewares/validators/task-requests.js:56]: Understand Thank you

File: test/unit/utils/rqlQueryParser.test.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      ],
+      key2: [{ operator: Operators.EXCLUDE, value: "value" }],
+    });
+  });
+  it("should parse sort queries", function () {
+    const queryString = "sort:key-desc";
+    const parser = new RQLQueryParser(queryString);
+    const sortQueries = parser.getSortQueries();
+    expect(sortQueries).to.deep.equal({
+      key: "desc",
```
Comment (shubhamsigdar1) [inline test/unit/utils/rqlQueryParser.test.js:42]: In object we expect others properties also 
 {
      operator: Operators.INCLUDE,
      value: value.trim(),
      type: QueryTypes.FILTER,
      key: key.trim(),
    };

**Review by shubhamsigdar1 — COMMENTED** [review]
great work @Ajeyakrishna-k, lots of learning. Thank you.

File: utils/RQLParser.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  operator: Operators;
+  key: string;
+  value: string;
+  type: QueryTypes;
+}
+
+interface FilterQueryValue {
+  operator: Operators;
+  value: string;
+}
```
Comment (sahsisunny) [inline utils/RQLParser.ts:21]: I think these types can be reused. Can we keep them in separate files?

File: test/unit/utils/rqlQueryParser.test.js
Comment (sahsisunny) [inline test/unit/utils/rqlQueryParser.test.js:1]: I noticed that you created a new file in TypeScript and another in JavaScript. Could you also create this file in TypeScript?

Comment (Ajeyakrishna-k) [reply test/unit/utils/rqlQueryParser.test.js:1]: Done.

File: utils/RQLParser.ts (hunk 5 above)
Comment (Ajeyakrishna-k) [reply utils/RQLParser.ts]: Done.

File: test/unit/utils/rqlQueryParser.test.js (hunk 4 above)
Comment (Ajeyakrishna-k) [reply test/unit/utils/rqlQueryParser.test.js:42]: No check the return type of `getSortQueries`

**Review by shubhamsigdar1 — APPROVED** [review]
LGTM

**Review by ardourApeX — APPROVED** [review]

**Review by sahsisunny — APPROVED** [review]

**Review by shubhamsigdar1 — APPROVED** [review]

**Pull request merged**